File: 1143-longest-common-subsequence/1143-longest-common-subsequence.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:

        dp = {}

        def backtrack(i, j):

            #Base Case
            if i >= len(text1) or j >= len(text2):
                return 0
            
            if (i, j) in dp:
                return dp[(i, j)]
            
            if text1[i] == text2[j]:
                result = 1 + backtrack(i + 1, j + 1)
            else:
                result = max(backtrack(i + 1, j), backtrack(i, j + 1))

            dp[(i, j)] = result
            return result
        
        return backtrack(0, 0)
        

File: 1143-longest-common-subsequence/1143-longest-common-subsequence.cpp
class Solution {
public:
    int longestCommonSubsequence(string text1, string text2) 
    {
        
        std::map<pair<int, int>, int> map;
        return dfs(0, 0, text1, text2, map);
    }

    int dfs(int i, int j, std::string& text1, std::string& text2, std::map<pair<int, int>, int>& map)
    {

        //Base Case
        if (i >= text1.size() || j >= text2.size())
        {
            return 0;
        }

        if (map.count({i, j}))
        {
            return map[{i, j}];
        }

        int result = 0;

        //Case 1. Increase both
        if (text1[i] == text2[j])
        {
            result = 1 + dfs(i + 1, j + 1, text1, text2, map);
        }
        else
        {

            //Case 2 incresae either!
            result = std::max(dfs(i + 1, j, text1, text2, map), dfs(i, j + 1, text1, text2, map));
        }

        //Add to dp
        map[{i, j}] = result;

        return result;
    }

private:

};